### PR TITLE
8300257: C2: vectorization fails on some simple Memory Segment loops

### PR DESCRIPTION
**Pull request by rwestrel**
In the test case `testByteLong1` (that's extracted from a memory
segment micro benchmark), the address of the store is initially:

```
(AddP#204 base#195 base#195 (AddL#164 (ConvI2L#158 (CastII#157 (LshiftI#107 iv#101))) invar#163))
```

(#numbers are node numbers to help the discussion).

`iv#101` is the `Phi` of a counted loop. `invar#163` is the
`baseOffset` load.

To eliminate the range check, the loop is transformed into a loop nest
and as a consequence the address above becomes:

```
(AddP#204 base#195 base#195 (AddL#164 (ConvI2L#158 (CastII#157 (LShiftI#107 (AddI#326 invar#308 iv#321)))) invar#163))
```

`invar#308` is some expression from a `Phi` of the outer loop.

That `AddP` is transformed multiple times to push the invariants out of loop:

```
(AddP#568 base#195 (AddP#556 base#195 base#195 invar#163) (ConvI2L#158 (CastII#157 (AddI#566 (LShiftI#565 iv#321) invar#577))))
```

then:

```
(AddP#568 base#195 (AddP#847 (AddP#556 base#195 base#195 invar#163) (AddL#838 (ConvI2L#793 (LShiftL#760 iv#767)) (ConvI2L#818 (CastII#779 invar#577)))))
```

and finally:

```
(AddP#568 base#195 (AddP#949 base#195 (AddP#855 base#195 (AddP#556 base#195 base#195 invar#163) (ConvI2L#818 (CastII#809 invar#577))) (ConvI2L#938 (LShiftI#896 iv#908))))
```

`AddP#855` is out of the inner loop. 

This doesn't vectorize because:

- there are 2 invariants in the address expression but superword only
  support one (tracked by `_invar` in `SWPointer`)

- there are more levels of `AddP` (4) than superword supports (3)

To fix that, I propose to no longer track the address elements in
`_invar`, `_negate_invar` and `_invar_scale` but instead to have a
single `_invar` which is an expression built by superword as it
follows chains of `addP` nodes. I kept the previous `_invar`,
`_negate_invar` and `_invar_scale` as debugging and use them to check
that what vectorized with the previous scheme still does.

I also propose lifting the restriction on 3 levels of `AddP` entirely.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300257](https://bugs.openjdk.org/browse/JDK-8300257): C2: vectorization fails on some simple Memory Segment loops

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [dbf08c38](https://git.openjdk.org/jdk/pull/12942/files/dbf08c38137362bb7273cd7120dd8908944e5581)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [35b94989](https://git.openjdk.org/jdk/pull/12942/files/35b9498914b96ac703f35fc6396f81f72c977650)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12942/head:pull/12942` \
`$ git checkout pull/12942`

Update a local copy of the PR: \
`$ git checkout pull/12942` \
`$ git pull https://git.openjdk.org/jdk.git pull/12942/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12942`

View PR using the GUI difftool: \
`$ git pr show -t 12942`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12942.diff">https://git.openjdk.org/jdk/pull/12942.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12942#issuecomment-1461799820)

### CODE REVIEW

**Comment by vnkozlov** [comment]
So the final expression should be seen by superword as next since `AddP#855` is invariant for inner loop:
```
(AddP#568 base#195 (AddP#949 base#195 (invar#855) (ConvI2L#938 (LShiftI#896 iv#908))))
```
First, I think you messed up with `()`. AddP node should have 3 inputs: base, address and offset. I don't see offset for `AddP#568`.
Second, if superword  can mark `AddP#855` as invariant (no need to parse it) your address expression become simple.

**Comment by rwestrel** [comment]
> So the final expression should be seen by superword as next since `AddP#855` is invariant for inner loop:
> 
> ```
> (AddP#568 base#195 (AddP#949 base#195 (invar#855) (ConvI2L#938 (LShiftI#896 iv#908))))
> ```
> 
> First, I think you messed up with `()`. AddP node should have 3 inputs: base, address and offset. I don't see offset for `AddP#568`. Second, if superword can mark `AddP#855` as invariant (no need to parse it) your address expression become simple.

I left the constant part of the address out of the expressions. So that was on purpose but confusing, sorry.
The current logic only sets the invariant to the result of an integer operation and otherwise follows through the AddP and checks that all of them have the same base. So If we wanted to use the `AddP` as invariant, we would need to change the logic so it stops as the first `AddP` that's loop invariant, we would likely not check that all `AddP` nodes have the same base and we would need to cast the `AddP` to an integer to compute the loop alignment.
Do you think that's better that what I'm proposing? I don't really have a strong opinion. Your suggestion is likely a smaller change.

**Comment by vnkozlov** [comment]
So we need to check all AddP to find base's offset. No short cuts then :(
I agree with your proposal with invariants. Last time I touched this code it was one of pain points that we have only one _invar. Any improvements to that is welcome.
I will start testing while looking on changes.

**Review by vnkozlov — APPROVED** [review]
My tier1-4,xcomp and stress testing passed.
I looked and changes and they seems fine.
May be we need to run performance testing too.

File: src/hotspot/share/opto/superword.hpp [hunk 1]
```diff
@@ -679,17 +682,17 @@ class SWPointer : public ArenaObj {
...
   // Comparable?
   bool invar_equals(SWPointer& q) {
-      return (_invar        == q._invar   &&
-              _invar_scale  == q._invar_scale &&
-              _negate_invar == q._negate_invar);
+    assert(_debug_invar == NodeSentinel || q._debug_invar == NodeSentinel ||
+           (_invar == q._invar) == (_debug_invar == q._debug_invar &&
+                                    _debug_invar_scale == q._debug_invar_scale &&
+                                    _debug_negate_invar == q._debug_negate_invar), "");
+    return _invar        == q._invar;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/superword.hpp:695]:
```suggestion
    return _invar == q._invar;
```

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMultiInvar.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Test
+    @IR(counts = { IRNode.LOAD_VECTOR, ">=1", IRNode.STORE_VECTOR, ">=1" })
+    public static void testLoopNest1(byte[] dest, byte[] src,
+                                    long start1, long stop1,
+                                    long start2, long stop2,
+                                    long start3, long stop3,
+                                    long start4, long stop4,
+                                    long start5, long stop5
+                                    ) {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMultiInvar.java:78]:
```suggestion
    public static void testLoopNest1(byte[] dest, byte[] src,
                                     long start1, long stop1,
                                     long start2, long stop2,
                                     long start3, long stop3,
                                     long start4, long stop4,
                                     long start5, long stop5) {
```

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMultiInvar.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Test
+    @IR(counts = { IRNode.LOAD_VECTOR, ">=1", IRNode.STORE_VECTOR, ">=1" })
+    public static void testLoopNest2(int[] dest, int[] src,
+                                    long start1, long stop1,
+                                    long start2, long stop2,
+                                    long start3, long stop3,
+                                    long start4, long stop4,
+                                    long start5, long stop5
+                                    ) {
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorizationMultiInvar.java:110]:
```suggestion
    public static void testLoopNest2(int[] dest, int[] src,
                                     long start1, long stop1,
                                     long start2, long stop2,
                                     long start3, long stop3,
                                     long start4, long stop4,
                                     long start5, long stop5) {
```

File: src/hotspot/share/opto/superword.cpp [hunk 4]
```diff
@@ -4313,41 +4308,34 @@ bool SWPointer::offset_plus_k(Node* n, bool negate) {
...
+  assert((_debug_invar == nullptr) == (_invar == nullptr), "");
 
   if (_analyze_only && is_loop_member(n)) {
     _nstack->push(n, _stack_idx++);
   }
   if (opc == Op_AddI) {
     if (n->in(2)->is_Con() && invariant(n->in(1))) {
-      _negate_invar = negate;
-      _invar = n->in(1);
+      maybe_add_to_invar(maybe_negate_invar(negate, n->in(1)));
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/superword.cpp:4318]: It feels like `maybe_negate_invar` should be moved into `maybe_add_to_invar` and be controlled by a `negate` argument.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I'll run some perf testing and report back.

**Comment by TobiHartmann** [comment]
Performance and correctness testing looks good.

File: src/hotspot/share/opto/superword.cpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/opto/superword.cpp]: Thanks for the review. I applied your suggestions and made that change too.

**Review by TobiHartmann — APPROVED** [review]
Thanks for making these changes. `SWPointer::maybe_negate_invar` could now be removed as it has only one user but I'm also fine with leaving it as is.

**Comment by rwestrel** [comment]
> Thanks for making these changes. `SWPointer::maybe_negate_invar` could now be removed as it has only one user but I'm also fine with leaving it as is.

Thanks for looking at this again.
Because it has a single use? The method is fairly simple but non trivial so I'll leave it as a separate method.

**Pull request closed without merging**